File: ParshinMaksim/Task3/Task3/MyTask.cs
namespace Task3;

public class MyTask<TResult> : IMyTask<TResult>
{
	private readonly Queue<Action> continuations = new();
	private readonly IMyThreadPool threadPool;
	private Func<TResult>? supplier;
	private TResult? result;
	private AggregateException? aggregateException;
	private readonly ManualResetEvent isResultReadyEvent = new(false);
	private readonly Object continuationQueueLocker = new();

	/// <summary>
	/// Returns true if the task was finished successfully or with exception
	/// </summary>
	public bool IsCompleted { get; private set; }

	/// <summary>
	/// Returns result of the task. If it has not been calculated, the current thread waits for it
	/// </summary>
	/// <exception cref="ThreadPoolShutdownException">Thrown if the thread pool was shutdown and result was not evaluated</exception>
	/// <exception cref="AggregateException">Thrown if another exception was thrown during task evaluation</exception>
	public TResult Result
	{
		get
		{
			var handles = new[] { threadPool.CancellationToken.WaitHandle, isResultReadyEvent };
			WaitHandle.WaitAny(handles);

			if (!IsCompleted && threadPool.CancellationToken.IsCancellationRequested)
			{
				throw new ThreadPoolShutdownException();
			}

			if (aggregateException != null)
			{
				throw aggregateException;
			}

			return result;
		}
	}

	private void StartTask()
	{
		try
		{
			result = supplier();
		}
		catch (Exception exception)
		{
			aggregateException = new AggregateException(exception);
		}
		finally
		{
			IsCompleted = true;
			supplier = null;
			isResultReadyEvent.Set();

			lock (continuationQueueLocker)
			{
				while (continuations.Count != 0)
				{
					threadPool.Enqueue(continuations.Dequeue());
				}
			}
		}
	}

	/// <summary>
	/// Enqueues the task execution to the task's thread pool
	/// </summary>
	public void Run()
	{
		threadPool.Enqueue(StartTask);
	}

	/// <summary>
	/// Creates a new task which is applied to the result of another one
	/// </summary>
	/// <typeparam name="TNewResult">New task result type</typeparam>
	/// <param name="newSupplier">Function to be encapsulated into the new task</param>
	/// <returns>Created task</returns>
	public IMyTask<TNewResult> ContinueWith<TNewResult>(Func<TResult, TNewResult> newSupplier)
	{
		var task = new MyTask<TNewResult>(() => newSupplier(Result), threadPool);

		lock (continuationQueueLocker)
		{
			if (IsCompleted)
			{
				threadPool.Enqueue(task.StartTask);
			}
			else
			{
				continuations.Enqueue(task.StartTask);
			}
		}

		return task;
	}

	public MyTask(Func<TResult> supplier, IMyThreadPool threadPool)
	{
		this.threadPool = threadPool;
		this.supplier = supplier;
	}
}

File: ErshovVladislav/Task3/src/main/java/ershov/mytask/MyTask.java
package main.java.ershov.mytask;

import main.java.ershov.threadpool.ThreadPool;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class MyTask<TResult> implements IMyTask<TResult> {

    private final Function<Void, TResult> function;
    private final ThreadPool threadPool;
    private volatile TResult resultTask = null;
    private volatile boolean isCompletedTask = false;
    private volatile String functionErrorMessage = null;
    private final List<Object> continuations = new ArrayList<>();

    public MyTask(Function<Void, TResult> function, ThreadPool threadPool) {
        this.function = function;
        this.threadPool = threadPool;
    }

    @Override
    public <TNewResult> IMyTask<TNewResult> continueWith(Function<TResult, TNewResult> function) {
        if (isCompleted() || getFunctionErrorMessage() != null) {
            throw new IllegalStateException("The original task has already been submitted for execution");
        }

        Function<Void, TResult> getResultFunction = input -> resultTask;
        IMyTask<TNewResult> newTask = new MyTask<>(getResultFunction.andThen(function), threadPool);

        if (isCompleted()) {
            threadPool.enqueue(newTask);
        } else {
            continuations.add(newTask);
        }

        return newTask;
    }

    @Override
    public boolean isCompleted() {
        return isCompletedTask;
    }

    @Override
    public String getFunctionErrorMessage() {
        return functionErrorMessage;
    }

    @Override
    public TResult result() throws InterruptedException, AggregateException {
        while (!isCompleted() || getFunctionErrorMessage() != null) {
            if (getFunctionErrorMessage() != null) {
                throw new AggregateException(getFunctionErrorMessage());
            }

            Thread.sleep(1000);
        }

        return resultTask;
    }

    @Override
    public void run() {
        if (getFunctionErrorMessage() != null || isCompleted()) {
            return;
        }

        try {
            resultTask = function.apply(null);
            isCompletedTask = true;
        } catch (Exception exception) {
            functionErrorMessage = exception.getMessage();

            for (Object task: continuations) {
                ((IMyTask<?>) task).setFunctionErrorMessage(exception.getMessage());
            }
        } finally {
            for (Object task: continuations) {
                threadPool.enqueue((IMyTask<?>) task);
            }
            System.out.println("Task finished with " + resultTask);
        }
    }

    @Override
    public void setFunctionErrorMessage(String message) {
        functionErrorMessage = message;
    }

}
